mrc-3318 Link data to model variables

EmmaLRussell opened the pull request:
This branch allows the user to link data columns to model variables. For Model Fit applications, after data has been uploaded and code compiled, when you go to the Options tab, you should see a new 'Link' section with a drop down list for each non-time column in the data, with all model variables available in the drop down lists. Selections from the drop down lists should perist across tab changes.  The user can, if they wish, select the same variabe to link to multiple columns, since only one link at a time will be used when the fit itself is executed.

The values selected will be updated when:
- new data is uploaded (since all columns may have changed)
- new time variable is selected in Data tab (since the old time variable should now be available to link, and the new time variable shouldn't)
- code is recompiled (since this could affect what model variables are available)

In all these cases, the app should retain any previous selections which can be retained e.g. on code change it will retain any selections of variables which still exist and remove any which do not.  

To test, run deps and app, browse to http://localhost:3000/apps/day2
Upload either of the attached data files (they have different columns so can be used for testing the case of uploading new data removing selections for old columns). 
Go to Options tab and test expected behaviour.  

To test the code change case, you can paste this code over the default code for the Day 2 app. This replaces the I variable with a J variable. Remember to compile the code before you'll see changes the linked variables.
```
# variables
deriv(S) <-  - beta * S * J / N
deriv(J) <- beta * S * J / N - sigma * J
deriv(R) <- sigma * J
# initial conditions
initial(S) <- N - I0
initial(J) <- I0
initial(R) <- 0
# parameters
N <- user(1e6)
I0 <- user(1)
beta <- user(4)
sigma <- user(2)
```

Other changes in this branch:
- file uploads are required to contain at least two columns (one time and one non-time)
- added an enum for app types

[influenza_data_MULTIPLE_TIME.csv](https://github.com/mrc-ide/wodin/files/9062649/influenza_data_MULTIPLE_TIME.csv)
[influenza_data_MULTIPLE_TIME_DIFFERENT.csv](https://github.com/mrc-ide/wodin/files/9062650/influenza_data_MULTIPLE_TIME_DIFFERENT.csv)

LekanAnni commented on app/static/src/app/components/options/LinkData.vue:
```diff
@@ -0,0 +1,60 @@
+        </div>
+        <div class="col-6">
+            <select class="form-select"
+                    @change="updateLinkedVariable(dataColumn, $event)" :value="linkedVariables[dataColumn] || ''">
```
```suggestion
                    @change="updateLinkedVariable(dataColumn, $event)"
                      :value="linkedVariables[dataColumn] || ''">
```
```suggestion
                    @change="updateLinkedVariable(dataColumn, $event)" :value="linkedVariables[dataColumn] || ''">
```

LekanAnni commented on app/static/src/app/components/options/LinkData.vue:
```diff
@@ -0,0 +1,60 @@
+
+        const updateLinkedVariable = (dataColumn: string, event: Event) => {
+            const selectValue = (event.target as HTMLSelectElement).value;
+            const value = selectValue === "" ? null : selectValue;
+            store.commit(`${namespace}/${FitDataMutation.SetLinkedVariable}`, { column: dataColumn, variable: value });
```
```suggestion
            store.commit(`${namespace}/${FitDataMutation.SetLinkedVariable}`, 
            { column: dataColumn, variable: selectValue || null });
```
```suggestion
            const value = selectValue === "" ? null : selectValue;
            store.commit(`${namespace}/${FitDataMutation.SetLinkedVariable}`, { column: dataColumn, variable: value });
```

LekanAnni reviewed (changes requested):
Works nicely, looks great, see my comments.

* apps/day2 does not really log apps behaviour on console. I find it very strange.
*  if an error occurred while compiling, the message render for link section on options tab displays `Please upload data and compile model in order to select links. ` regardless of wether a file is uploaded to not. There should be a different message when code is invalid but file is uploaded.
* I noticed that file upload does not re-upload same file twice here as well. I think the browser checks if the file has changed, if not then it will not re-upload.

Nicolas-Dolan reviewed (approved):
Looks good to me

EmmaLRussell commented:
>    apps/day2 does not really log apps behaviour on console. I find it very strange.

I've added the mutation logging plugin to work in model fir apps

>    if an error occurred while compiling, the message render for link section on options tab displays Please upload data and compile model in order to select links. regardless of wether a file is uploaded to not. There should be a different message when code is invalid but file is uploaded.

I've changed this so that the component now displays relevant message(s) only

>    I noticed that file upload does not re-upload same file twice here as well. I think the browser checks if the file has changed, if not then it will not re-upload.

Ha, good point! I've made it clear the current value when user first clicks the button. That means we can upload a newer version of the previous file, but we can still see the filename after we've done so.

LekanAnni reviewed (approved):
LGTM 👍 

Thanks
